### PR TITLE
schannel: Disable ALPN on Windows < 8.1

### LINKED ISSUE #840: "schannel: failed to retrieve ALPN result" when running on Windows 7
### I did this
We built libcurl with Visual Studio 2015 on Windows Server 2012 R2 (based on Windows 8.1) and linked it statically to an app that downloads a file via HTTP/1.1. This app was then executed on Windows 7. It failed with this error when trying to download [this file](https://www.kullo.net/download/files/windows/latest-all):

> schannel: failed to retrieve ALPN result

(see also #724, where ALPN support for schannel has been introduced)

### Suspected root cause
The error occurs when libcurl calls [QueryContextAttributes](https://msdn.microsoft.com/en-us/library/windows/desktop/aa379340%28v=vs.85%29.aspx) in [schannel.c](https://github.com/curl/curl/blob/master/lib/vtls/schannel.c#L634). This function has a minimum supported OS version of Windows 8.1 or Server 2012 R2. It is not guaranteed to work on older versions, like Windows 7 (where it fails when called with `SECPKG_ATTR_APPLICATION_PROTOCOL`).

### I expected the following
The download should have succeeded.

### curl/libcurl version
7.49.1 (all is well with 7.48.0)

### operating system
* libcurl was built on Windows Server 2012 R2
* the build was used on Windows 7

### PR DESCRIPTION
**Pull request by tiwoc**
Calling QueryContextAttributes with SECPKG_ATTR_APPLICATION_PROTOCOL fails on Windows < 8.1 (and < Server 2012 R2), so we need to disable ALPN on these OS versions.

Fixes #840

I built curl on Windows Server 2012 R2 with MSVS 2015 and the command line `nmake /f Makefile.vc MODE=dll VC=14 DEBUG=no`. I tested the result with `curl.exe -v https://google.com/` on both the build machine and a Windows 7 box.

TL;DR: The behavior on Windows Server 2012 R2 doesn't change, but on Windows 7, the patch fixes the TLS connection by disabling ALPN.

### Windows Server 2012 R2, master
```
*   Trying 216.58.214.110...
* Connected to google.com (216.58.214.110) port 443 (#0)
* schannel: SSL/TLS connection with google.com port 443 (step 1/3)
* schannel: checking server certificate revocation
* schannel: ALPN, offering http/1.1
* schannel: sending initial handshake data: sending 198 bytes...
[...]
* schannel: SSL/TLS connection with google.com port 443 (step 3/3)
* schannel: ALPN, server accepted to use http/1.1
* schannel: incremented credential handle refcount = 1
```

### Windows Server 2012 R2, patched
Same result as with master, as expected.

### Windows 7, master
```
*   Trying 216.58.201.238...
* Connected to google.com (216.58.201.238) port 443 (#0)
* schannel: SSL/TLS connection with google.com port 443 (step 1/3)
* schannel: checking server certificate revocation
* schannel: ALPN, offering http/1.1
* schannel: sending initial handshake data: sending 179 bytes...
[...]
* schannel: SSL/TLS connection with google.com port 443 (step 3/3)
* schannel: failed to retrieve ALPN result
* Closing connection 0
* schannel: shutting down SSL/TLS connection with google.com port 443
* schannel: clear security context handle
* schannel: clear credential handle
curl: (35) schannel: failed to retrieve ALPN result
```

### Windows 7, patched
```
*   Trying 216.58.201.238...
* Connected to google.com (216.58.201.238) port 443 (#0)
* schannel: SSL/TLS connection with google.com port 443 (step 1/3)
* schannel: checking server certificate revocation
* schannel: sending initial handshake data: sending 179 bytes...
[...]
* schannel: SSL/TLS connection with google.com port 443 (step 3/3)
* schannel: incremented credential handle refcount = 1
```

### CODE REVIEW

**Comment by bagder** [comment]
That patch seems to cause compiler problems: https://ci.appveyor.com/project/curlorg/curl/build/7.47.0.1070/job/ooasinfy9qlsu6gp#L2460

**Comment by tiwoc** [comment]
I just updated the commit to hopefully make it compatible to older MS compilers.

**Comment by captain-caveman2k** [comment]
I've pushed a fix based on your patch @tiwoc that uses the newly added Curl_verify_windows_version() function instead of IsWindows8Point1OrGreater().

Whilst the ALPN code is dependent on the v8.1 SDK where this function is present in VersionHelpers.h the function simply uses VerifyVersionInfo() underneath just like Curl_verify_windows_version() does. Not only that but older SDKs may not include VersionHelpers.h.

Please also note that when curl runs under Windows 8.1 and Windows 10, version 6.2 (aka Windows 8.0 / Windows Server 2012) will be reported if the application's manifest does not defined 8.1 or 10 as the target OS :(
